sdk/core/Azure.Core/src/Shared/OperationInternalBase.cs

```diff
@@ -176,14 +190,38 @@ public virtual Response WaitForCompletionResponse(CancellationToken cancellation
         /// <param name="cancellationToken">A <see cref="CancellationToken"/> controlling the request lifetime.</param>
         /// <returns>The last HTTP response received from the server, including the final result of the long-running operation.</returns>
         /// <exception cref="RequestFailedException">Thrown if there's been any issues during the connection, or if the operation has completed with failures.</exception>
-        public virtual Response WaitForCompletionResponse(TimeSpan pollingInterval, CancellationToken cancellationToken)
+        public Response WaitForCompletionResponse(TimeSpan pollingInterval, CancellationToken cancellationToken)
         {
             OperationPoller poller = new OperationPoller(_fallbackStrategy);
             return poller.WaitForCompletionResponse(UpdateStatus, () => HasCompleted, () => RawResponse, pollingInterval, cancellationToken);
         }
 
+        public bool TrySetPendingResponse(Response response, CancellationToken cancellationToken)
+            => TrySetPendingResponseAsync(false, response, cancellationToken).EnsureCompleted();
+
+        public async ValueTask<bool> TrySetPendingResponseAsync(Response response, CancellationToken cancellationToken)
+            => await TrySetPendingResponseAsync(true, response, cancellationToken).ConfigureAwait(false);
+
+        private async ValueTask<bool> TrySetPendingResponseAsync(bool async, Response response, CancellationToken cancellationToken)
+        {
+            using var asyncLock = await _operationStateLock.GetLockOrValueAsync(async, cancellationToken).ConfigureAwait(false);
+            if (asyncLock.HasValue)
+            {
+                return false;
+            }
+
+            RawResponse = response;
+            return true;
+        }
+
         private async ValueTask<Response> UpdateStatusAsync(bool async, CancellationToken cancellationToken)
         {
+            using var asyncLock = await _operationStateLock.GetLockOrValueAsync(async, cancellationToken).ConfigureAwait(false);
```

Comment:
Here is where synchronization of calls happens.

Reply:
Same here. Would help with future maintenance.